Notes/Algorithms/Data Structures/Hash-Tables.c

```diff
@@ -1,230 +1,244 @@
+        *success = SUCCESS;
         return item;
     }
+    return NULL;
 }
 
-// insert takes in a key and value
-// calls a helper function to create an Item struct with the provided key and value
-// adds the new Item to the hashTable
-// if the expected value is in the hashTable at the expected key then returns a success code
-void insert(int type, int key, int value, int success) {
-    if (type == TEST) {
-        testHash[key] = createItem(key, value, success);
-        if (testHash[key]->value != value) {
-            success = ERROR;
-        }
-        
-    } else if (type == EXPECTED) {
-        expectedHash[key] = createItem(key, value, success);
-        if (expectedHash[key]->value != value) {
-            success = ERROR;
-        }
+// creates a new HashTable struct
+// returns new HashTable struct
+HashTable * createHashTable(int size) {
+    if (size < 1 || size > 100) return NULL;
+    HashTable * hashTable = malloc(sizeof(HashTable));
+    hashTable->items = malloc(sizeof(Item) * size);
+    if (hashTable == NULL || hashTable->items == NULL) {
+        return NULL;
     }
+    hashTable->size = size;
+    for (int i = 0; i < size; i++) {
+        hashTable->items[i] = NULL;
+    }
+    return hashTable;
 }
 
-int testInsert() {
-    return ERROR;
-}
 
-void buildInsertTests() {
-    //InsertTests
-    // 5 passes
-    // 5 Fails
-    // Empty Array
-    // Too large for hash table
-    // Collision
-    // Full Hash Table
+// insert 
+// adds the new Item struct to the hashTable
+// if the expected value is in the hashTable at the expected key then returns a success code
+int insert(HashTable * hashTable, char key[], int value) {
+    int success = ERROR;
+    int index = hashing(hashTable, key);
+    if (isCollision(hashTable, index)) {
+        //find next free space  
+        while(hashTable->items[index] != NULL && hashTable->items[index]->key != key && hashTable->items[index]->key != -1) 
+        { 
+            index++; 
+            index % hashTable->size; 
+        } 
+    }
+    hashTable->items[index] = createItem(key, value, &success);
+    if (hashTable->items[index]->value != value) {
+        return ERROR;
+    }
+    return SUCCESS;
 }
-//To-do: will need to think of how to handle collisions 
-// buildHashTable receives an array of keys and an array of values
-// key[i] will be mapped to value[i] when building the hashTable
-// loops through the values to create a key for the current value then inserts it into the existing hashTable.
+
+// bulkInsert 
 // returns a SUCCESS code if all values are added into the hashTable successfully 
-int buildHashTable(int type, int *keys, int *values) {
-    int errcode = SUCCESS;
-    for (int i = 0; i < SIZE; i++) {
-        int key = hashing(keys[i]);
-        insert(type, key, values[i], errcode);
-   }
-   return errcode;
+// returns an ERROR code if a value cannot be added into the hashTable 
+int bulkInsert(HashTable * hashTable, char keys[][10], int* values) {
```

Comment:
Why does `keys[][10]` have a fixed size for the 2nd dimension?

Some checking to make sure that `hashTable`, `keys`, and `values` are all the same cardinality would be good defensive programming.

Reply:
I made the fixed size after running into issues passing a 2d char array to a function and found this solution:
https://www.geeksforgeeks.org/pass-2d-array-parameter-c/
In our conversation yesterday I completely forgot that I had looked into this before but without the fixed size I run into seg faults because of the 2d array parameter.

Defensive programming added: [commit 15ae918eee495055daee78cb452cba8c007eb601](https://github.com/kcarrel/Training-Notes/pull/12/commits/15ae918eee495055daee78cb452cba8c007eb601). Will expand on testing for this with the insert tests I'll be working on next commit.